ir/base.def

```diff
@@ -318,64 +311,83 @@ class Annotation {
-// FIXME -- get rid of this class -- classes with annotations should have an
-// inline Vector<Annotation> instead (remove useless indirection)
-class Annotations {
-    optional inline Vector<Annotation> annotations;
-    Annotations { if (!srcInfo && !annotations.empty()) srcInfo = annotations[0]->srcInfo; }
-    static Annotations *empty;  // FIXME -- should be const
-    size_t size() const { return annotations.size(); }
-    // Get the annotation with the specified name or nullptr.
-    // There should be at most one annotation with this name.
-    Annotation getSingle(cstring name) const { return get(annotations, name); }
-    Annotations add(Annotation annot) {
-        if (annot->srcInfo) srcInfo += annot->srcInfo;
-        annotations.push_back(annot);
-        return this; }
-    Annotations add(Annotation annot) const { return clone()->add(annot); }
-    Annotations addAnnotation(cstring name, Expression expr, bool structured = false) {
-        return add(new Annotation(name, Vector<Expression>(expr), structured)); }
-    Annotations addAnnotation(cstring name, Expression expr, bool structured = false) const {
-        return add(new Annotation(name, Vector<Expression>(expr), structured)); }
-    // Add annotation if another annotation with the same name is not
-    // already present.
-    Annotations addAnnotationIfNew(cstring name, Expression expr, bool structured = false) const {
-        return getSingle(name) ? this : addAnnotation(name, expr, structured); }
-    Annotations append(Annotations other) {
-        if (other == nullptr)
-            return this;
-        for (auto a : other->annotations)
-            add(a);
-        return this;
-    }
-    // If annotations with the same name are already present, remove them.
-    // Add this annotation.
-    Annotations addOrReplace(cstring name, Expression expr, bool structured = false) const {
-        auto rv = clone();
-        remove_if(rv->annotations, [name](const Annotation *a)->bool { return a->name == name; });
-        rv->annotations.push_back(new Annotation(name, Vector<Expression>(expr), structured));
-        return rv; }
+    /// An unparsed annotation body
+    inline Vector<AnnotationToken> body;
+
+    /// Annotations that are simple expressions
+    inline Vector<Expression> expr;
+
+    /// Annotations described as key-value pairs
+    inline IndexedVector<NamedExpression> kv;
+
 #emit
-    typedef std::function<bool(const IR::Annotation*)> Filter;
+    /// Whether the annotation body needs to be parsed.
+    /// Invariant: if this is true, then expr and kv must both be empty. If the
+    /// annotation is compiler-generated (e.g., derived from a P4₁₄ pragma),
+    /// then needsParsing will be false and the body will be empty, but expr or
+    /// kv may be populated.
+    bool needsParsing : 1;
+
+    /// If this is true this is a structured annotation, and there are some
+    /// constraints on its contents.
+    bool structured : 1;
```

Comment:
Does the bit-field really decrease size of the struct? Since there are the `Vector`/`IndexedVector` members, I would expect the alignment to be 64bit and therefore it would not really meatter if the 2 bools are 2 bits or 2 bytes. Or is there another reason to use bit-field?

Reply:
I am planning to fix this in subsequent PRs :) We really need a discriminated union here as `sizeof(Annotation`) is terrible large. For now I just put it here not to forget.